### ISSUE TITLE
libbpf 1.0: deprecate or make more uniform bpf_prog_info_linear-related APIs

### ISSUE THREAD

**Issue by anakryiko** [libbpf/libbpf#313]
Details: https://github.com/libbpf/libbpf/wiki/Libbpf:-the-road-to-v1.0#libbpfh-high-level-apis

**Comment by davemarchevsky** [comment]
Detailed comments from the road to 1.0 wiki:

> #313 bpf_prog_info_linear-related APIs. They completely fail libbpf naming guidelines (they use bpf_program__ prefix, but they don’t operate on bpf_program objects). bpf_program__get_prog_info_linear() doesn’t even declare that it is expecting enum bpf_prog_info_array and just goes with generic \_\_u64. They seem to be only used in perf and bpftool, so updating all users seems doable. At the very least, I’d fix the naming to be libbpf_ prefixed. But I need input from Song, Arnaldo and others here.

Talked to @anakryiko and @liu-song-6 to get some rough consensus on what should be done here. Andrii's main points were:
  * Is this doing something so usecase-specific that it should be pulled out into the few usecases that need it?
  * If answer to the above is "no", can it be made more obvious - either via renaming, refactor, or improved porcelain wrappers - that it's generally useful?

My opinion as someone fairly new to libbpf internals is that `get_prog_info_linear` is generally useful so the latter approach should be taken. If anyone wants to get the variable-len arrays that `bpf_prog_get_info_by_fd` provides they'd need to implement the same double `bpf_obj_get_info_by_fd` call pattern that `get_prog_info_linear` is doing, which feels like unintuitive yak shaving that shouldn't be duplicated. It should be renamed to something that better indicates that it actually can grab _all_ the `bpf_prog_info` data, variable-len or not, so that this is more obvious. Also name should follow libbpf naming guidelines, perhaps some small refactor or added user-friendly porcelain as well. 

All the variable-len stuff should still be shoved into one buffer to keep existing `perf` and `bpftool` uses happy. I suspect that any breaking changes are unlikely to break out-of-tree usecases - I don't think the function is being used outside `perf`, `bpftool` as it's not obvious that it's generally useful. Of all the BPF-adjacent stuff at FB, I found a single user of the variable-len `bpf_prog_info` data, and they reimplemented the double `bpf_obj_get_info_by_fd` call pattern instead of using `get_prog_info_linear`. Search for out-of-tree public code using `_linear` call was not fruitful.

@liu-song-6 was fine with either "pull out into few usecases that need it" or "rename/refactor", main concern with the latter being not breaking `perf`/`bpftool`. @anakryiko wanted to look at the code a bit more but was supportive of the suggested approach.

@acmel @arcamelo do you have an opinion here? Notes mention consulting you as well.

**Comment by anakryiko** [comment]
Ok, looking a bit more at this I think we should kill this (i.e., deprecate and move to bpftool/perf). It looks very-very perf-specific. There are extra bpf_program__bpil_addr_to_offs() and bpf_program__bpil_offs_to_addr() APIs just to be able to work with bpf_prog_info_liner, which is already a smell.

As for your concern about yak shaving by calling bpf_prog_get_info_by_fd() twice. This is not the only API that has this convention for variable-sized data that kernel can provide where user-space might not know the size beforehand. BTF info and BPF link info fetching has similar conventions. It's impractical to provide some higher-level APIs just to save few lines of code for users. Also these APIs are informational/diagnostic, so they aren't used by everyone and even if they are, you need to understand the API protocol to be able to use it anyways.

The whole requirement of having the memory in a single continuously allocated piece of memory looks rather arbitrary and again optimized for perf needs. I don't see how typical users would have such requirement.

So in short, I think low-level API is pretty straightforward to use as is, tbh. Two calls, one to learn the lenghts with NULL pointers, then allocate memory (one array for each kind of data you need or one contiguous block of memory) is up to user to manage (maybe they want to reuse memory between calls to multiple BPF programs, that would be more efficient for tools like bpftool, btw. Then final call to populate that user-owned memory. Then it can free or reuse it, it's none of libbpf concerns.

So I'd say let's deprecate it. Bpftool has 3 use cases for this, two of them only fetch one piece of data, another fetches 5, but fixed 5. So it's not a problem to implement this. And moreso, we can actually make it more efficient on bpftool side by reusing memory, but that's an optional benefit.

**Comment by davemarchevsky** [comment]
> There are extra bpf_program__bpil_addr_to_offs() and bpf_program__bpil_offs_to_addr() APIs just to be able to work with bpf_prog_info_liner, which is already a smell.

I believe these are specific to `perf`'s use of the data - persisting and reading from file - and would be necessary regardless of the data being "linear" or not. Totally agree that these should be sent to `perf`'s codebase.

>  Also these APIs are informational/diagnostic, so they aren't used by everyone and even if they are, you need to understand the API protocol to be able to use it anyways.

> The whole requirement of having the memory in a single continuously allocated piece of memory looks rather arbitrary and again optimized for perf needs. I don't see how typical users would have such requirement.

> So in short, I think low-level API is pretty straightforward to use as is, tbh.

Most of this is seems like a matter of taste and assumptions of "average libbpf user". You've certainly talked to more average libbpf users than me so I'll defer to your judgement here.

**Comment by anakryiko** [comment]
It's obviously a judgement call, but it seems like a right decision.

We probably will get back to providing a bit better API around obj_get_info API as a separate task, though. So not all is lost, don't worry :)

But in the current incarnation get_prog_info_linear should go.
